File: Design-Patterns-Using-Ruby/Adapter/AdapterPatternRectangle.rb
# Desired Interface
module Shape
    def draw
        raise NotImplementedError, 'Subclasses must implement this method'
    end
end

# Circle class
class Circle include Shape
    def initialize(x, y, radius)
        @x = x
        @y = y
        @radius = radius
    end

    def draw
        puts 'Draw the Circle.'
    end
end

# Rectangle class (Adaptee)
class Rectangle
    def initialize(x, y, length, width)
        @x = x
        @y = y
        @length = length
        @width = width
    end

    def old_draw
        puts 'Drawing Rectangle.'
    end
end

# RectangleAdapter class
class RectangleAdapter include Shape
    def initialize(x, y, length, width)
        @adaptee = Rectangle.new(x, y, length, width)
    end

    def draw
        @adaptee.old_draw
    end
end
  
# Client Code
adapter = RectangleAdapter.new(1, 2, 3, 4)
adapter.draw

=begin 
Drawing Rectangle.
=end

File: Design-Patterns-Using-CSharp/Adapter/AdapterPatternRectangle.cs
using System;

// Desired Interface
interface Shape
{
    void Draw();
}

// Circle class
class Circle : Shape
{
    private int x, y, radius;

    public Circle(int x, int y, int r)
    {
        this.x = x;
        this.y = y;
        this.radius = r;
    }

    public void Draw()
    {
        Console.WriteLine("Draw the Circle.");
    }
}

// Rectangle class (Adaptee)
class Rectangle
{
    private int x, y, length, width;

    public Rectangle(int x, int y, int l, int w)
    {
        this.x = x;
        this.y = y;
        this.length = l;
        this.width = w;
    }

    public void OldDraw()
    {
        Console.WriteLine("Drawing Rectangle.");
    }
}

// RectangleAdapter class
class RectangleAdapter : Shape
{
    private Rectangle adaptee;

    public RectangleAdapter(int x, int y, int l, int w)
    {
        this.adaptee = new Rectangle(x, y, l, w);
    }

    public void Draw()
    {
        adaptee.OldDraw();
    }
}

// Client Code
public class AdapterPatternRectangle
{
    public static void Main(string[] args)
    {
        Shape adapter = new RectangleAdapter(1, 2, 3, 4);
        adapter.Draw();
    }
}

/* 
Drawing Rectangle.
*/
